Make strict schema

duncandewhurst opened the pull request:
Related to #1046

Depends on https://github.com/open-contracting/ocdskit/pull/184

Related issues from 1.2.0 Strict milestone:

- [x] #881
- [x] #859
- [x] #814
- [x] #679
- [x] #718
- [x] #669
- [x] #650

TO-DO:

- [x] Remove deprecated types
- [x] Decide how to keep the strict schema up to date with changes to the regular schema
- [x] Update strict schema to match 1.2-dev
- [ ] ~~Update requirements.txt once https://github.com/open-contracting/ocdskit/pull/184 is released on PyPI~~
- [ ] ~~Add documentation~~ (To be done in a separate PR)
- [x] Update changelog
- [x] Create a compiled-release-schema.json and remove "null" types and null enums from all fields.

duncandewhurst committed:
Run ocdskit schema-strict

duncandewhurst committed:
Add additional schema validation (#881)

duncandewhurst committed:
Make id, name required on Organization[Reference] (#679)

duncandewhurst committed:
Make id required on Amendment and RelatedProcess (650)

duncandewhurst committed:
Run manage.py pre-commit

duncandewhurst committed:
Move strict schema files to subfolder

duncandewhurst committed:
Checkout schema files from 1.2-dev

duncandewhurst committed:
Update derivative strict schema files

duncandewhurst committed:
Merge branch '1.2-dev' into 1046-strict-schema

duncandewhurst commented:
>  Decide how to keep the strict schema up to date with changes to the regular schema

@jpmckinney - so far I made the changes in this PR manually, but in order to keep the strict schema up to date, I can add a script as part of `manage.py pre-commit` to derive the strict schema programmatically. Does that sound good?

jpmckinney commented:
Yes, let's do it that way. Then we can run the script and see if the result is currently the same as your manual preparation :)

duncandewhurst committed:
Generate strict schema files programmatically

duncandewhurst committed:
Fix Flake8 errors

duncandewhurst committed:
strict schema: remove integer type from id fields

duncandewhurst committed:
strict schema: remove null types from package schemas

duncandewhurst commented:
Unfortunately, I'd made the manual changes to an old version of the schema and since merged the latest version of 1.2-dev so it's easier to just to diff the regular and strict schemas to check that the changes are as expected:

```bash
git diff --word-diff=color --no-index schema/release-schema.json schema/strict/release-schema.json
git diff --word-diff=color --no-index schema/release-package-schema.json schema/strict/release-package-schema.json
git diff --word-diff=color --no-index schema/record-package-schema.json schema/strict/record-package-schema.json
```

duncandewhurst commented:
The tests are failing on an unrelated issue from the organization units example and on the schema file for the new `UnitValue` definition (suggested in https://github.com/open-contracting/standard/issues/881). I'm not sure if there's a better way to add that programmatically. Hardcoding it in `manage.py` seemed like a bad idea.

jpmckinney commented:
> I'm not sure if there's a better way to add that programmatically.

Add what? The UnitValue definition? I figure we can add it to the regular schema (without extra validation keywords), and then just add the validation keywords in the strict schema.

duncandewhurst committed:
Move key-based validation properties logic from OCDS Kit

duncandewhurst committed:
Move UnitValue definition to regular release schema

duncandewhurst committed:
Fix flake8 errors

duncandewhurst committed:
Run ./manage.py pre-commit

duncandewhurst committed:
Strict schema: Remove integer type from projectID

duncandewhurst commented:
> > I'm not sure if there's a better way to add that programmatically.
> 
> Add what? The UnitValue definition? I figure we can add it to the regular schema (without extra validation keywords), and then just add the validation keywords in the strict schema.

I've added the UnitValue definition to the regular schema and updated `manage.py` to set the `$ref` for `Unit.Value` in the strict schema.

Do you have a view on if and how we should feature the strict schema(s) in the documentation?

I figure few users would do anything with the actual JSON schema, but adding schema viewers for the strict versions of the release, release package and record package schemas seems overly duplicative. I think we should include some mention of it in the docs if we are going to include it in the DRT and helpdesk feedback processes. We could either add an admonition to https://standard.open-contracting.org/staging/1.2-dev/en/schema/# or a new page under the reference section which can give a summary of the differences to the regular schema and direct users to the DRT to check their data against the strict schema.

jpmckinney commented:
> Do you have a view on if and how we should feature the strict schema(s) in the documentation?

I think a concise admonition as you suggest would be best. It should briefly describe why the strict schema is preferred for new implementations (and how it can be useful to improve data quality of existing implementations).

In terms of duplication, we can consider using https://sphinx-panels.readthedocs.io/en/latest/ to allow browsing of each schema. That way it's not taking up a lot of the page.

duncandewhurst commented:
> In terms of duplication, we can consider using https://sphinx-panels.readthedocs.io/en/latest/ to allow browsing of each schema. That way it's not taking up a lot of the page.

I tried this approach but only the first docson widget is displayed. The second widget throws an Uncaught Error: only one instance of babel-polyfill is allowed (see https://github.com/lbovet/docson/issues/73).

In other projects that have multiple docson widgets per page we use https://github.com/OpenDataServices/docson which is a fork of an earlier version of docson that doesn't use babel-polyfill, but I see that https://github.com/open-contracting/docson makes several improvements that would be lost if we used the other fork.

I found https://www.npmjs.com/package/idempotent-babel-polyfill/ which might be a solution but my Javascript knowledge falls short of how to update docson to use it.

duncandewhurst commented:
@jpmckinney based on my previous comment, do you still want to use Sphinx Panels?

jpmckinney commented:
> In other projects that have multiple docson widgets per page we use https://github.com/OpenDataServices/docson which is a fork of an earlier version of docson that doesn't use babel-polyfill

Are you sure? I looked at several branches at https://github.com/OpenDataServices/docson/branches, comparing them to the source repository, and I don't see any removal of babel-polyfill.

Anyway, let's do this in different PRs:

1. Make the changes for `manage.py` and `UnitValue`.
2. Start drafting the changes to the documentation. When ready to merge, run the `manage.py` command. As part of this, I can sort out removing the polyfill, which I think is not necessary for today's browsers.

duncandewhurst committed:
Merge branch '1.2-dev' into 1046-strict-schema

Add missing type to UnitValue.currency. Run ./manage.py pre-commit.

duncandewhurst committed:
Merge branch '1.2-dev' into 1046-strict-schema

duncandewhurst committed:
Update changelog

duncandewhurst commented:
> > In other projects that have multiple docson widgets per page we use https://github.com/OpenDataServices/docson which is a fork of an earlier version of docson that doesn't use babel-polyfill
> 
> Are you sure? I looked at several branches at https://github.com/OpenDataServices/docson/branches, comparing them to the source repository, and I don't see any removal of babel-polyfill.

The ODS fork doesn't remove babel-polyfill. Rather, babel-polyfill was added to the original docson repo after the ODS repo was forked and before the open-contracting repo was forked.

In https://github.com/OpenDataServices/docson/compare/master...lbovet:docson:master there is a commit from the original repo that adds babel-polyfill: https://github.com/OpenDataServices/docson/commit/0756db7653257b9d438bb0536932a8020ae46f66, which also appears in the open-contracting fork: https://github.com/open-contracting/docson/commit/0756db7653257b9d438bb0536932a8020ae46f66

> Anyway, let's do this in different PRs:
> 
>     1. Make the changes for `manage.py` and `UnitValue`.

These changes were already made in this PR: https://github.com/open-contracting/standard/pull/1480/commits/85af27bee511efb3a1b0c04e016a333aa067ae15. Is that OK?

>     2. Start drafting the changes to the documentation. When ready to merge, run the `manage.py` command. As part of this, I can sort out removing the polyfill, which I think is not necessary for today's browsers.

So create a new PR for the documentation changes?

jpmckinney commented:
Noting that sphinx-panels is obsolete. Need to use https://github.com/executablebooks/sphinx-design.

jpmckinney commented:
To clarify my earlier request in https://github.com/open-contracting/standard/pull/1480#issuecomment-1198822301

Let's try to get something merged sooner than later.

To do that, let's not generate the `strict` files in this PR. You can remove those files, and I'll "squash and merge". That means this PR is part 1 (manage.py and UnitValue).

In a new PR, we can add the documentation (the last checklist item). Once the docs are ready, we can run the `manage.py` command to generate the `strict` files, and then merge.

jpmckinney committed:
Merge branch '1.2-dev' into 1046-strict-schema

jpmckinney commented:
Note to myself that I haven't actually reviewed the changes yet, as I'm not sure if the PR is only "draft" due to the missing documentation changes.

jpmckinney committed:
manage: Update strict code to use merge_props=True

jpmckinney committed:
chore: flake8

duncandewhurst committed:
Merge branch '1.2-dev' into 1046-strict-schema

duncandewhurst committed:
manage.py: Sort imports

duncandewhurst committed:
schema: Remove strict files

duncandewhurst committed:
Update changelog, reinstate whitespace

duncandewhurst commented:
> Let's try to get something merged sooner than later.

Erm... I'm not sure how I missed this one, sorry! It's now ready for review. I included a changelog entry for `UnitValue` even though it makes no difference to users. Feel free to remove, or we can just update it once the part 2 PR is done.

duncandewhurst committed:
Merge branch '1.2-dev' into 1046-strict-schema

duncandewhurst committed:
manage.py: Update pre-commit to reflect changes in json_dump/json_load

duncandewhurst committed:
manage.py (pre-commit): Make strict record schema

duncandewhurst commented:
I've updated this PR to reflect the record schema being moved into its own file and to reflect the changes made to `manage.py` on `1.2-dev`.

Edit: I also temporarily removed `ruff-format` from `.pre-commit-config.yaml`. It can be reinstated once https://github.com/open-contracting/standard/pull/1708#issuecomment-2433776370 is done.

duncandewhurst committed:
Merge branch '1.2-dev' into 1046-strict-schema

duncandewhurst committed:
.pre-commit-config.yaml: Reinstate ruff-format

odscjen commented:
Added a new to-do from https://github.com/open-contracting/standard/issues/330#issuecomment-2450280930

duncandewhurst commented:
Have moved back to draft whilst I make those changes.

duncandewhurst committed:
manage.py (pre-commit): Create a compiled release schema, format code

duncandewhurst committed:
manage.py (pre-commit): Update references to strict schemas
